src/paas_charm/charm_state.py

```diff
@@ -338,10 +205,85 @@ class IntegrationsState:
 
     redis_uri: str | None = None
     databases_uris: dict[str, str] = field(default_factory=dict)
-    s3_parameters: S3Parameters | None = None
-    saml_parameters: SamlParameters | None = None
+    s3_parameters: "S3Parameters | None" = None
+    saml_parameters: "SamlParameters | None" = None
     rabbitmq_uri: str | None = None
-    tempo_parameters: TempoParameters | None = None
+    tempo_parameters: "TempoParameters | None" = None
+
+    @classmethod
+    def generate_saml_relation_parameters(
+        cls,
+        saml_relation_data: typing.MutableMapping[str, str] | None,
+        parameter_type: type,
+    ) -> "SamlParameters | None":
+        """Generate SAML relation parameter class from relation data.
+
+        Args:
+            saml_relation_data: Relation data.
+            parameter_type: Parameter type to use.
+
+        Return:
+            Parameter instance created.
+
+        Raises:
+            CharmConfigInvalidError: If some parameter in invalid.
+        """
+        if saml_relation_data is not None:
+            try:
+                relation_parameter = parameter_type(**saml_relation_data)
+            except ValidationError as exc:
+                error_message = build_validation_error_message(exc)
+                raise CharmConfigInvalidError(f"Invalid configuration: {error_message}") from exc
+        else:
+            relation_parameter = None
+        return relation_parameter
+
+    @classmethod
+    def generate_relation_parameters(
+        cls,
+        relation_data: dict[str, str] | None,
+        parameter_type: type,
+    ) -> "SamlParameters | S3Parameters | TempoParameters | None":
+        """Generate relation parameter class from relation data.
+
+        Args:
+            relation_data: Relation data.
+            parameter_type: Parameter type to use.
+
+        Return:
+            Parameter instance created.
+
+        Raises:
+            CharmConfigInvalidError: If some parameter in invalid.
+        """
+        if relation_data:
+            try:
+                relation_parameter = parameter_type(**relation_data)
+            except ValidationError as exc:
+                error_message = build_validation_error_message(exc)
+                raise CharmConfigInvalidError(f"Invalid configuration: {error_message}") from exc
```

Comment:
Since the error message becomes generic we should make sure that the error message now doesn't degrade user experience